src/components/TreeReactFlow/index.tsx

```diff
@@ -529,143 +637,446 @@ const makePrimerNode = async (
+  const sigEdgeId = "def-sig-" + def.name.baseName;
+  const bodyEdgeId = "def-body-" + def.name.baseName;
+  const defNameNode: PrimerNodeWithNestedAndDef = {
+    id: defNodeId,
+    data: {
+      def: def.name,
+      width: p.nodeWidth * p.nameNodeMultipliers.width,
+      height: p.nodeHeight * p.nameNodeMultipliers.height,
+      selected: deepEqualTyped(p.selection, {
+        tag: "SelectionDef",
+        contents: { def: def.name },
+      }),
+      nested: [],
+    },
+    type: "primer-def-name",
+    zIndex: 0,
+  };
+  const defEdge = async (
+    tree: APITree,
+    nodeType: NodeType,
+    edgeId: string
+  ): Promise<[Tree<PrimerNodeWithNestedAndDef, PrimerEdge>, PrimerEdge]> =>
+    augmentTree(tree, (n0) =>
+      makePrimerNode(
+        n0,
+        p,
+        p.layout,
+        0,
+        { tag: "termDefNode", nodeType },
+        def.name
+      ).then(([n, e, nested]) => [
+        primerNodeWith(n, {
+          def: def.name,
+          nested: nested.map((g) =>
+            graphMap(g, ({ position, ...n }) => ({
+              ...primerNodeWith(n, { def: def.name }),
+              position,
+            }))
+          ),
+        }),
+        e,
+      ])
+    ).then((t) => [
+      t,
+      {
+        id: edgeId,
+        source: defNodeId,
+        target: t.node.id,
+        type: "primer-def",
+        zIndex: 0,
+        sourceHandle: Position.Bottom,
+        targetHandle: Position.Top,
+      },
+    ]);
+  const sigTree = await defEdge(def.type_, "SigNode", sigEdgeId);
+  const bodyTree = await (def.term
+    ? defEdge(def.term, "BodyNode", bodyEdgeId)
+    : undefined);
+  return {
+    node: defNameNode,
+    childTrees: [sigTree, ...(bodyTree ? [bodyTree] : [])],
+  };
+};
+
+const typeDefToTree = async (
+  def: TypeDef,
+  p: DefParams & NodeParams & { layout: LayoutParams }
+): Promise<Tree<PrimerNode<{ def: GlobalName }>, PrimerEdge>> => {
+  type N = PrimerNode<{ def: GlobalName }>;
+  type E = PrimerEdge;
+  type T = Tree<N, E>;
+
+  const rootId = "typedef-name-" + def.name.baseName;
+  const paramsTree = def.params.reduceRight<
+    [T, (parentId: string) => E] | undefined
+  >((child, name) => {
+    const id = "typedef-param-" + name;
+    const node: N = {
+      id,
+      type: "primer-typedef-param",
+      data: {
+        def: def.name,
+        width: p.nodeWidth,
+        height: p.nodeHeight,
+        name,
+        selected: deepEqualTyped(p.selection, {
+          tag: "SelectionTypeDef",
+          contents: {
+            def: def.name,
+            node: {
+              tag: "TypeDefParamNodeSelection",
+              contents: name,
+            },
+          },
+        }),
+      },
+      zIndex: 0,
+    };
+    return [
+      {
+        node,
+        childTrees: [],
+        ...(child
+          ? { rightChild: mapSnd((f: (parentId: string) => E) => f(id))(child) }
+          : {}),
+      },
+      (parentId) => ({
+        id: JSON.stringify([parentId, id]),
```

Comment:
We should factor out using `JSON.stringify([source,target])` for edge IDs. This was duplicated a bit before this PR, but it's now a lot worse.